### PR TITLE
Validate the serialization/deserialization of composed types in typescript 

### PR DESCRIPTION
**Pull request by koros**
Add unit tests to validate the logic in the serialization/deserialization of composed types

### CODE REVIEW

**Review by andrueastman — APPROVED** [review]
👍🏼

File: packages/serialization/json/test/common/jsonSerializationWriter.ts [hunk 1]
```diff
@@ -230,4 +230,51 @@ describe("JsonParseNode", () => {
 		const contentAsStr = decoder.decode(serializedContent);
 		assert.equal(contentAsStr, '{"id":"1","title":"title","location":{"address":{"city":"Redmond","postalCode":"98052","state":"Washington","street":"NE 36th St"},"coordinates":{"latitude":47.678581,"longitude":-122.131577},"displayName":"Microsoft Building 25","floorCount":50,"hasReception":true,"contact":null},"keywords":[{"created":"2023-07-26T10:41:26Z","label":"Keyword1","termGuid":"10e9cc83-b5a4-4c8d-8dab-4ada1252dd70","wssId":6442450941},{"created":"2023-07-26T10:51:26Z","label":"Keyword2","termGuid":"2cae6c6a-9bb8-4a78-afff-81b88e735fef","wssId":6442450942}],"extra":{"value":{"createdDateTime":{"value":"2024-01-15T00:00:00+00:00"}}}}');
 	});
+
+	it("it should serialize a union of object and primitive when the value is a string", async () => {
+		const inputObject: TestParser = {
```
Comment (baywet) [inline packages/serialization/json/test/common/jsonSerializationWriter.ts:235]: I'd also like to see additional tests for unions of primitives and object types (e.g. `string | User`)

**Review by baywet — CHANGES_REQUESTED** [review]
Great initiative! I think we should extend the test cases here.

File: packages/serialization/json/test/common/testEntity.ts [hunk 2]
```diff
@@ -7,6 +7,8 @@
 
 import type { BackedModel, BackingStore, Parsable, ParseNode, SerializationWriter } from "@microsoft/kiota-abstractions";
 import { Guid } from "guid-typescript";
+import { write } from "node:fs";
+import test from "node:test";
```
Comment (baywet) [inline packages/serialization/json/test/common/testEntity.ts:11]:
```suggestion
```

**Review by baywet — CHANGES_REQUESTED** [review]

File: packages/serialization/json/test/common/unionOfObjectsAndPrimitives.ts [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+export function serializePetGetResponse(writer: SerializationWriter, petGetResponse: Partial<PetGetResponse> | undefined = {}): void {
+	switch (typeof petGetResponse.data) {
+		case "number":
+			writer.writeNumberValue("data", petGetResponse.data);
+			break;
+		case "string":
+			writer.writeStringValue("data", petGetResponse.data);
+			break;
+		default:
+			writer.writeObjectValue<Cat | Dog | number | string>("data", petGetResponse.data, serializePetGetResponse_data);
```
Comment (baywet) [inline packages/serialization/json/test/common/unionOfObjectsAndPrimitives.ts:203]: shouldn't the generic type parameter here contain only the object types?

Comment (rkodev) [reply packages/serialization/json/test/common/unionOfObjectsAndPrimitives.ts:203]: This is resolved in the latest generation.

Comment (baywet) [reply packages/serialization/json/test/common/unionOfObjectsAndPrimitives.ts:203]: Can you push an update please? and resolve the conflicts as well?

**Review by baywet — APPROVED** [review]
Thank you for making the changes!

**Pull request merged**